From 687197490c22b27e0a32414e7f6fc5592ec88372 Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Fri, 16 Aug 2019 11:12:10 +0800
Subject: [PATCH] Add support for posix_devctl()

It is defined in POSIX 1003.26 and required by the FACE Technical
Standard.

POSIX 1003.26 is available from here:
https://standards.ieee.org/findstds/standard/1003.26-2003.html

Implement FACE 3.0 specific details, as documented in Section 3.2.1.2, GPSS
POSIX numbers 10 and 11, see:
http://oaevents.goprecise.com/wp-content/uploads/2017/11/FACE-Technical-Standard-Edition-3.0.pdf

Upstream-Status: Pending

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 bits/confname.h                               |   2 +
 posix/Makefile                                |   3 +-
 posix/Versions                                |   5 +
 posix/devctl.h                                |  51 ++++++++
 posix/getconf.c                               |   2 +
 posix/posix_devctl.c                          |  31 +++++
 posix/sysconf.c                               |   7 ++
 posix/unistd.h                                |   4 +
 sysdeps/mach/hurd/i386/libc.abilist           |   1 +
 sysdeps/unix/sysv/linux/aarch64/libc.abilist  |   1 +
 sysdeps/unix/sysv/linux/alpha/libc.abilist    |   1 +
 sysdeps/unix/sysv/linux/arm/be/libc.abilist   |   1 +
 sysdeps/unix/sysv/linux/arm/le/libc.abilist   |   1 +
 sysdeps/unix/sysv/linux/csky/libc.abilist     |   1 +
 sysdeps/unix/sysv/linux/hppa/libc.abilist     |   1 +
 sysdeps/unix/sysv/linux/i386/libc.abilist     |   1 +
 sysdeps/unix/sysv/linux/ia64/libc.abilist     |   1 +
 .../sysv/linux/m68k/coldfire/libc.abilist     |   1 +
 .../unix/sysv/linux/m68k/m680x0/libc.abilist  |   1 +
 .../sysv/linux/microblaze/be/libc.abilist     |   1 +
 .../sysv/linux/microblaze/le/libc.abilist     |   1 +
 .../sysv/linux/mips/mips32/fpu/libc.abilist   |   1 +
 .../sysv/linux/mips/mips32/nofpu/libc.abilist |   1 +
 .../sysv/linux/mips/mips64/n32/libc.abilist   |   1 +
 .../sysv/linux/mips/mips64/n64/libc.abilist   |   1 +
 sysdeps/unix/sysv/linux/nios2/libc.abilist    |   1 +
 sysdeps/unix/sysv/linux/posix_devctl.c        | 116 ++++++++++++++++++
 .../linux/powerpc/powerpc32/fpu/libc.abilist  |   1 +
 .../powerpc/powerpc32/nofpu/libc.abilist      |   1 +
 .../linux/powerpc/powerpc64/be/libc.abilist   |   1 +
 .../linux/powerpc/powerpc64/le/libc.abilist   |   1 +
 .../unix/sysv/linux/riscv/rv64/libc.abilist   |   1 +
 .../unix/sysv/linux/s390/s390-32/libc.abilist |   1 +
 .../unix/sysv/linux/s390/s390-64/libc.abilist |   1 +
 sysdeps/unix/sysv/linux/sh/be/libc.abilist    |   1 +
 sysdeps/unix/sysv/linux/sh/le/libc.abilist    |   1 +
 .../sysv/linux/sparc/sparc32/libc.abilist     |   1 +
 .../sysv/linux/sparc/sparc64/libc.abilist     |   1 +
 .../unix/sysv/linux/x86_64/64/libc.abilist    |   1 +
 .../unix/sysv/linux/x86_64/x32/libc.abilist   |   1 +
 40 files changed, 251 insertions(+), 1 deletion(-)
 create mode 100644 posix/devctl.h
 create mode 100644 posix/posix_devctl.c
 create mode 100644 sysdeps/unix/sysv/linux/posix_devctl.c

diff --git a/bits/confname.h b/bits/confname.h
index 5dc8215093..e287b44153 100644
--- a/bits/confname.h
+++ b/bits/confname.h
@@ -492,6 +492,8 @@ enum
 #define _SC_LEVEL4_CACHE_ASSOC		_SC_LEVEL4_CACHE_ASSOC
     _SC_LEVEL4_CACHE_LINESIZE,
 #define _SC_LEVEL4_CACHE_LINESIZE	_SC_LEVEL4_CACHE_LINESIZE
+    _SC_POSIX_26_VERSION,
+#define _SC_POSIX_26_VERSION		_SC_POSIX_26_VERSION
     /* Leave room here, maybe we need a few more cache levels some day.  */
 
     _SC_IPV6 = _SC_LEVEL1_ICACHE_SIZE + 50,
diff --git a/posix/Makefile b/posix/Makefile
index 4507d84bf1..965d333022 100644
--- a/posix/Makefile
+++ b/posix/Makefile
@@ -23,7 +23,7 @@ subdir	:= posix
 include ../Makeconfig
 
 headers	:= sys/utsname.h sys/times.h sys/wait.h sys/types.h unistd.h	      \
-	   glob.h regex.h wordexp.h fnmatch.h 				      \
+	   glob.h regex.h wordexp.h fnmatch.h devctl.h			      \
 	   getopt.h bits/getopt_core.h bits/getopt_ext.h bits/getopt_posix.h  \
 	   bits/types.h bits/typesizes.h bits/pthreadtypes.h		      \
 	   bits/pthreadtypes-arch.h bits/thread-shared-types.h		      \
@@ -64,6 +64,7 @@ routines :=								      \
 	spawnattr_getsigmask spawnattr_getschedpolicy spawnattr_getschedparam \
 	spawnattr_setsigmask spawnattr_setschedpolicy spawnattr_setschedparam \
 	posix_madvise							      \
+	posix_devctl							      \
 	get_child_max sched_cpucount sched_cpualloc sched_cpufree \
 	streams-compat
 
diff --git a/posix/Versions b/posix/Versions
index 7d06a6d0c0..0208d96514 100644
--- a/posix/Versions
+++ b/posix/Versions
@@ -147,6 +147,11 @@ libc {
   }
   GLIBC_2.30 {
   }
+
+  GLIBC_2.28 {
+    posix_devctl;
+  }
+
   GLIBC_PRIVATE {
     __libc_fork; __libc_pread; __libc_pwrite;
     __nanosleep_nocancel; __pause_nocancel;
diff --git a/posix/devctl.h b/posix/devctl.h
new file mode 100644
index 0000000000..1010035a26
--- /dev/null
+++ b/posix/devctl.h
@@ -0,0 +1,51 @@
+/* Definitions for POSIX posix_devctl interface.
+   Copyright (C) 2014-2019 Wind River Systems, Inc.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>. */
+
+#ifndef _DEVCTL_H
+#define _DEVCTL_H
+
+#include <unistd.h>
+/*
+ * posix_devctl() is defined by POSIX 1003.26-2003. It requires
+ * POSIX_26_C_SOURCE to be defined in order for the function to
+ * be available.
+ */
+#if _POSIX_26_C_SOURCE == 200312L
+
+#include <sys/ioctl.h>
+#include <sys/types.h>
+
+/* http://oaevents.goprecise.com/wp-content/uploads/2017/11/FACE-Technical-Standard-Edition-3.0.pdf
+ * Implement FACE 3.0 specific details, as documented in Section 3.2.1.2, GPSS
+ * POSIX numbers 10 and 11, see:
+ * An OSS UoC shall define:
+ * a. A macro in <sys/ioctl.h> named FIONBIO to configure sockets for non-blocking I/O
+ * b. A macro in <devctl.h> named SOCKCLOSE to close an open socket
+ * Use `_IOR' to define SOCKCLOSE that avoids macro conflict with <sys/ioctl.h>
+ */
+#define SOCKCLOSE    _IOR('f', 200, int)
+
+__BEGIN_DECLS
+extern int posix_devctl(int fildes,
+                        int dcmd,
+                        void *__restrict dev_data_ptr,
+                        size_t nbyte,
+                        int *__restrict dev_info_ptr);
+__END_DECLS
+
+#endif /* _POSIX_26_C_SOURCE == 200312L */
+#endif /* _DEVCTL_H */
diff --git a/posix/getconf.c b/posix/getconf.c
index a8ba876c14..ffd88feb14 100644
--- a/posix/getconf.c
+++ b/posix/getconf.c
@@ -403,6 +403,8 @@ static const struct conf vars[] =
     { "_POSIX_IPV6", _SC_IPV6, SYSCONF },
     { "_POSIX_RAW_SOCKETS", _SC_RAW_SOCKETS, SYSCONF },
 
+    { "_POSIX_26_VERSION", _SC_POSIX_26_VERSION, SYSCONF },
+
     { NULL, 0, SYSCONF }
   };
 
diff --git a/posix/posix_devctl.c b/posix/posix_devctl.c
new file mode 100644
index 0000000000..a002274bf0
--- /dev/null
+++ b/posix/posix_devctl.c
@@ -0,0 +1,31 @@
+/* Copyright (C) 2019 Wind River Systems, Inc.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>. */
+
+#include <devctl.h>
+#include <errno.h>
+
+/* The posix_devctl method is defined by POSIX 1003.26-2003.
+   The interface can simply pass the arguments to the regular system ioctl. */
+
+int posix_devctl(int fildes,
+                 int dcmd,
+                 void *__restrict dev_data_ptr,
+                 size_t nbyte,
+                 int *__restrict dev_info_ptr)
+{
+    return ENOSYS;
+}
+stub_warning (posix_devctl)
diff --git a/posix/sysconf.c b/posix/sysconf.c
index ba303384c1..68b677a944 100644
--- a/posix/sysconf.c
+++ b/posix/sysconf.c
@@ -148,6 +148,13 @@ __sysconf (int name)
     case _SC_GETPW_R_SIZE_MAX:
       return NSS_BUFLEN_PASSWD;
 
+    case _SC_POSIX_26_VERSION:
+#ifdef _POSIX_26_C_SOURCE
+      return _POSIX_26_VERSION;
+#else
+      return -1;
+#endif
+
     case _SC_ARG_MAX:
     case _SC_CHILD_MAX:
     case _SC_CLK_TCK:
diff --git a/posix/unistd.h b/posix/unistd.h
index 0dd4200ad6..26ad94923d 100644
--- a/posix/unistd.h
+++ b/posix/unistd.h
@@ -85,6 +85,10 @@ __BEGIN_DECLS
    creation of locales with the localedef utility.  */
 #define _POSIX2_LOCALEDEF       __POSIX2_THIS_VERSION
 
+/* If defined, the implementation supports the
+   implement posix_devctl() using ioctl()  */
+#define _POSIX_26_C_SOURCE 200312L
+
 /* X/Open version number to which the library conforms.  It is selectable.  */
 #ifdef __USE_XOPEN2K8
 # define _XOPEN_VERSION	700
diff --git a/sysdeps/mach/hurd/i386/libc.abilist b/sysdeps/mach/hurd/i386/libc.abilist
index 7c2cb2b05a..ab70a47ad6 100644
--- a/sysdeps/mach/hurd/i386/libc.abilist
+++ b/sysdeps/mach/hurd/i386/libc.abilist
@@ -2042,6 +2042,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.29 _hurd_port_move F
 GLIBC_2.29 posix_spawn_file_actions_addchdir_np F
diff --git a/sysdeps/unix/sysv/linux/aarch64/libc.abilist b/sysdeps/unix/sysv/linux/aarch64/libc.abilist
index a4c31932cb..0251083be6 100644
--- a/sysdeps/unix/sysv/linux/aarch64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/aarch64/libc.abilist
@@ -2133,6 +2133,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/alpha/libc.abilist b/sysdeps/unix/sysv/linux/alpha/libc.abilist
index e7f2174ac2..9ad7c4ba22 100644
--- a/sysdeps/unix/sysv/linux/alpha/libc.abilist
+++ b/sysdeps/unix/sysv/linux/alpha/libc.abilist
@@ -2033,6 +2033,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/arm/be/libc.abilist b/sysdeps/unix/sysv/linux/arm/be/libc.abilist
index b152c0e24a..8751c88481 100644
--- a/sysdeps/unix/sysv/linux/arm/be/libc.abilist
+++ b/sysdeps/unix/sysv/linux/arm/be/libc.abilist
@@ -118,6 +118,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/arm/le/libc.abilist b/sysdeps/unix/sysv/linux/arm/le/libc.abilist
index 9371927927..6cac372b3b 100644
--- a/sysdeps/unix/sysv/linux/arm/le/libc.abilist
+++ b/sysdeps/unix/sysv/linux/arm/le/libc.abilist
@@ -118,6 +118,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/csky/libc.abilist b/sysdeps/unix/sysv/linux/csky/libc.abilist
index 9b3cee65bb..215c099602 100644
--- a/sysdeps/unix/sysv/linux/csky/libc.abilist
+++ b/sysdeps/unix/sysv/linux/csky/libc.abilist
@@ -1525,6 +1525,7 @@ GLIBC_2.29 remque F
 GLIBC_2.29 rename F
 GLIBC_2.29 renameat F
 GLIBC_2.29 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.29 revoke F
 GLIBC_2.29 rewind F
 GLIBC_2.29 rewinddir F
diff --git a/sysdeps/unix/sysv/linux/hppa/libc.abilist b/sysdeps/unix/sysv/linux/hppa/libc.abilist
index df6d96fbae..27bca7ca9e 100644
--- a/sysdeps/unix/sysv/linux/hppa/libc.abilist
+++ b/sysdeps/unix/sysv/linux/hppa/libc.abilist
@@ -1880,6 +1880,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/i386/libc.abilist b/sysdeps/unix/sysv/linux/i386/libc.abilist
index fcb625b6bf..5f59fb4b44 100644
--- a/sysdeps/unix/sysv/linux/i386/libc.abilist
+++ b/sysdeps/unix/sysv/linux/i386/libc.abilist
@@ -2045,6 +2045,7 @@ GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/ia64/libc.abilist b/sysdeps/unix/sysv/linux/ia64/libc.abilist
index cb556c5998..176b6bcead 100644
--- a/sysdeps/unix/sysv/linux/ia64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/ia64/libc.abilist
@@ -1914,6 +1914,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/m68k/coldfire/libc.abilist b/sysdeps/unix/sysv/linux/m68k/coldfire/libc.abilist
index 5e3cdea246..499b0dea0e 100644
--- a/sysdeps/unix/sysv/linux/m68k/coldfire/libc.abilist
+++ b/sysdeps/unix/sysv/linux/m68k/coldfire/libc.abilist
@@ -119,6 +119,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/m68k/m680x0/libc.abilist b/sysdeps/unix/sysv/linux/m68k/m680x0/libc.abilist
index ea5e7a41af..bd08bb2881 100644
--- a/sysdeps/unix/sysv/linux/m68k/m680x0/libc.abilist
+++ b/sysdeps/unix/sysv/linux/m68k/m680x0/libc.abilist
@@ -1989,6 +1989,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/microblaze/be/libc.abilist b/sysdeps/unix/sysv/linux/microblaze/be/libc.abilist
index ac55b0acd7..58aa710617 100644
--- a/sysdeps/unix/sysv/linux/microblaze/be/libc.abilist
+++ b/sysdeps/unix/sysv/linux/microblaze/be/libc.abilist
@@ -2125,6 +2125,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/microblaze/le/libc.abilist b/sysdeps/unix/sysv/linux/microblaze/le/libc.abilist
index f7ced487f7..f39b6e3616 100644
--- a/sysdeps/unix/sysv/linux/microblaze/le/libc.abilist
+++ b/sysdeps/unix/sysv/linux/microblaze/le/libc.abilist
@@ -2125,6 +2125,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/mips/mips32/fpu/libc.abilist b/sysdeps/unix/sysv/linux/mips/mips32/fpu/libc.abilist
index 06c2e64edd..fac8900e1b 100644
--- a/sysdeps/unix/sysv/linux/mips/mips32/fpu/libc.abilist
+++ b/sysdeps/unix/sysv/linux/mips/mips32/fpu/libc.abilist
@@ -1967,6 +1967,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/mips/mips32/nofpu/libc.abilist b/sysdeps/unix/sysv/linux/mips/mips32/nofpu/libc.abilist
index bdfd073b86..72a1dddd7f 100644
--- a/sysdeps/unix/sysv/linux/mips/mips32/nofpu/libc.abilist
+++ b/sysdeps/unix/sysv/linux/mips/mips32/nofpu/libc.abilist
@@ -1965,6 +1965,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/mips/mips64/n32/libc.abilist b/sysdeps/unix/sysv/linux/mips/mips64/n32/libc.abilist
index 3d61d4974a..a3ed8d9cf1 100644
--- a/sysdeps/unix/sysv/linux/mips/mips64/n32/libc.abilist
+++ b/sysdeps/unix/sysv/linux/mips/mips64/n32/libc.abilist
@@ -1973,6 +1973,7 @@ GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/mips/mips64/n64/libc.abilist b/sysdeps/unix/sysv/linux/mips/mips64/n64/libc.abilist
index 675acca5db..2f8759f4d4 100644
--- a/sysdeps/unix/sysv/linux/mips/mips64/n64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/mips/mips64/n64/libc.abilist
@@ -1968,6 +1968,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/nios2/libc.abilist b/sysdeps/unix/sysv/linux/nios2/libc.abilist
index 7fec0c9670..e7fef8aa60 100644
--- a/sysdeps/unix/sysv/linux/nios2/libc.abilist
+++ b/sysdeps/unix/sysv/linux/nios2/libc.abilist
@@ -2166,6 +2166,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/posix_devctl.c b/sysdeps/unix/sysv/linux/posix_devctl.c
new file mode 100644
index 0000000000..2fd7ae61a8
--- /dev/null
+++ b/sysdeps/unix/sysv/linux/posix_devctl.c
@@ -0,0 +1,116 @@
+/* Define the posix_devctl.c function.
+   Copyright (C) 2014-2019 Wind River Systems, Inc.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>. */
+
+#include <devctl.h>
+#include <errno.h>
+
+/* The posix_devctl method is defined by POSIX 1003.26-2003.
+   The interface can simply pass the arguments to the regular system ioctl. */
+
+int posix_devctl(int fildes,
+                 int dcmd,
+                 void *__restrict dev_data_ptr,
+                 size_t nbyte,
+                 int *__restrict dev_info_ptr)
+{
+  /* POSIX 1003.26-2003 indicates that there is a "general goal
+   * of being able to implement posix_devctl() using the existing
+   * ioctl() interfaces provided in current UNIX systems and other
+   * POSIX implementations because the standard allows but does not
+   * require checking the size of the device data."
+   *
+   * As such this function will be implemented primarily using the
+   * existing ioctl() interface as permitted by the specification.
+   *
+   * Additionally the FACE Technical Standard 3.0, Section 3.2.1.2,
+   * requiret that FIONBIO and SOCKCLOSE command be supported.
+   */
+
+  int ret = 0;
+  __set_errno (0);
+
+  if (fildes < 0)
+  {
+    __set_errno (EBADF);
+    ret = -1;
+    goto on_error;
+  }
+
+  if (nbyte > 0 && dev_data_ptr == NULL)
+  {
+    __set_errno (EINVAL);
+    ret = -1;
+    goto on_error;
+  }
+
+  /* An OSS UoC shall provide:
+   * a. The SOCKCLOSE command in the posix_devctl() API
+   * b. The FIONBIO command in the posix_devctl() API
+   */
+  switch (dcmd)
+    {
+    case SOCKCLOSE:
+      if (nbyte > 0 || dev_data_ptr != NULL)
+      {
+        __set_errno (EINVAL);
+        ret = -1;
+        goto on_error;
+      }
+      ret = close(fildes);
+      break;
+    // FIONBIO is implemented in kernel
+    case FIONBIO:
+      if ((nbyte > 0 && nbyte < sizeof(int)) ||
+          dev_data_ptr == NULL)
+      {
+        __set_errno (EINVAL);
+        ret = -1;
+        goto on_error;
+      }
+    default:
+      ret = __ioctl(fildes, dcmd, dev_data_ptr);
+    }
+
+
+on_error:
+  /* ioctl() and posix_devctl() have different return value semantics.
+   *
+   * (from the Linux man-pages project)
+   * ioctl() will usually return success with zero, but a few ioctl() request use
+   * the return value as an output parameter and return a nonnegative value on
+   * success.  On error, -1 is returned, and errno is set appropriately.
+   *
+   * post_devctl(), upon successful completion, shall return zero; otherwise an
+   * error number shall be returned to indicate the error. The value returned via
+   * the dev_info_ptr argument is driver dependent.
+   *
+   * Based on this it can be assumed an ioctl() return value of -1 indicates
+   * an error, and the error number will be set in errno.
+   *
+   * Any nonnegative value returned is then considered to indicate success,
+   * and assumed to be an output parameter from ioctl().  Output parameters are
+   * transferred into dev_info_ptr, unless it is NULL.
+   */
+  if (dev_info_ptr != NULL) {
+      *dev_info_ptr = ret;
+  }
+
+  if (ret < 0)
+    ret = errno;
+
+  return ret;
+}
diff --git a/sysdeps/unix/sysv/linux/powerpc/powerpc32/fpu/libc.abilist b/sysdeps/unix/sysv/linux/powerpc/powerpc32/fpu/libc.abilist
index 1e8ff6f83e..72c65fdb6c 100644
--- a/sysdeps/unix/sysv/linux/powerpc/powerpc32/fpu/libc.abilist
+++ b/sysdeps/unix/sysv/linux/powerpc/powerpc32/fpu/libc.abilist
@@ -1993,6 +1993,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/powerpc/powerpc32/nofpu/libc.abilist b/sysdeps/unix/sysv/linux/powerpc/powerpc32/nofpu/libc.abilist
index b5a0751d90..7ee4d1f6c2 100644
--- a/sysdeps/unix/sysv/linux/powerpc/powerpc32/nofpu/libc.abilist
+++ b/sysdeps/unix/sysv/linux/powerpc/powerpc32/nofpu/libc.abilist
@@ -1997,6 +1997,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/powerpc/powerpc64/be/libc.abilist b/sysdeps/unix/sysv/linux/powerpc/powerpc64/be/libc.abilist
index 0c86217fc6..03ddd806ed 100644
--- a/sysdeps/unix/sysv/linux/powerpc/powerpc64/be/libc.abilist
+++ b/sysdeps/unix/sysv/linux/powerpc/powerpc64/be/libc.abilist
@@ -118,6 +118,7 @@ GLIBC_2.27 wcstof64 F
 GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/powerpc/powerpc64/le/libc.abilist b/sysdeps/unix/sysv/linux/powerpc/powerpc64/le/libc.abilist
index 2229a1dcc0..87c2cca687 100644
--- a/sysdeps/unix/sysv/linux/powerpc/powerpc64/le/libc.abilist
+++ b/sysdeps/unix/sysv/linux/powerpc/powerpc64/le/libc.abilist
@@ -2223,6 +2223,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/riscv/rv64/libc.abilist b/sysdeps/unix/sysv/linux/riscv/rv64/libc.abilist
index 31010e6cf7..71d893c9bb 100644
--- a/sysdeps/unix/sysv/linux/riscv/rv64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/riscv/rv64/libc.abilist
@@ -2095,6 +2095,7 @@ GLIBC_2.27 xprt_register F
 GLIBC_2.27 xprt_unregister F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/s390/s390-32/libc.abilist b/sysdeps/unix/sysv/linux/s390/s390-32/libc.abilist
index 4feca641b0..2b1591d808 100644
--- a/sysdeps/unix/sysv/linux/s390/s390-32/libc.abilist
+++ b/sysdeps/unix/sysv/linux/s390/s390-32/libc.abilist
@@ -2002,6 +2002,7 @@ GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/s390/s390-64/libc.abilist b/sysdeps/unix/sysv/linux/s390/s390-64/libc.abilist
index efe588a072..b4cc3bb658 100644
--- a/sysdeps/unix/sysv/linux/s390/s390-64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/s390/s390-64/libc.abilist
@@ -1907,6 +1907,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/sh/be/libc.abilist b/sysdeps/unix/sysv/linux/sh/be/libc.abilist
index 6bfc2b7439..21618eaf77 100644
--- a/sysdeps/unix/sysv/linux/sh/be/libc.abilist
+++ b/sysdeps/unix/sysv/linux/sh/be/libc.abilist
@@ -1884,6 +1884,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/sh/le/libc.abilist b/sysdeps/unix/sysv/linux/sh/le/libc.abilist
index 4b057bf4a2..4cece847be 100644
--- a/sysdeps/unix/sysv/linux/sh/le/libc.abilist
+++ b/sysdeps/unix/sysv/linux/sh/le/libc.abilist
@@ -1884,6 +1884,7 @@ GLIBC_2.27 wcstof64_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/sparc/sparc32/libc.abilist b/sysdeps/unix/sysv/linux/sparc/sparc32/libc.abilist
index 49cd597fd6..66bdef24a5 100644
--- a/sysdeps/unix/sysv/linux/sparc/sparc32/libc.abilist
+++ b/sysdeps/unix/sysv/linux/sparc/sparc32/libc.abilist
@@ -1996,6 +1996,7 @@ GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/sparc/sparc64/libc.abilist b/sysdeps/unix/sysv/linux/sparc/sparc64/libc.abilist
index 95e68e0ba1..1682176aec 100644
--- a/sysdeps/unix/sysv/linux/sparc/sparc64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/sparc/sparc64/libc.abilist
@@ -1937,6 +1937,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist b/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
index 1f2dbd1451..b643e8563e 100644
--- a/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
@@ -1895,6 +1895,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
diff --git a/sysdeps/unix/sysv/linux/x86_64/x32/libc.abilist b/sysdeps/unix/sysv/linux/x86_64/x32/libc.abilist
index 59da85a5d8..acf5a152f3 100644
--- a/sysdeps/unix/sysv/linux/x86_64/x32/libc.abilist
+++ b/sysdeps/unix/sysv/linux/x86_64/x32/libc.abilist
@@ -2146,6 +2146,7 @@ GLIBC_2.27 wcstof64x F
 GLIBC_2.27 wcstof64x_l F
 GLIBC_2.28 fcntl64 F
 GLIBC_2.28 renameat2 F
+GLIBC_2.28 posix_devctl F
 GLIBC_2.28 statx F
 GLIBC_2.28 thrd_current F
 GLIBC_2.28 thrd_equal F
-- 
2.18.2

